lib/puppetserver/ca/action/prune.rb

```diff
@@ -0,0 +1,149 @@
+require 'optparse'
+require 'openssl'
+require 'puppetserver/ca/errors'
+require 'puppetserver/ca/utils/cli_parsing'
+require 'puppetserver/ca/utils/file_system'
+require 'puppetserver/ca/utils/config'
+require 'puppetserver/ca/x509_loader'
+
+module Puppetserver
+  module Ca
+    module Action
+      class Prune
+        include Puppetserver::Ca::Utils
+
+        SUMMARY = "Prune the local CRL on disk to remove any duplicated certificates"
+        BANNER = <<-BANNER
+Usage:
+  puppetserver ca prune [--help]
+  puppetserver ca prune [--config]
+
+Description:
+  Prune the list of revoked certificates of any duplication within it.  This command
+  will only prune the CRLs that was signed by Puppet.
+
+Options:
+BANNER
+
+        def initialize(logger)
+          @logger = logger
+        end
+
+        def self.parser(parsed = {})
+          OptionParser.new do |opts|
+            opts.banner = BANNER
+            opts.on('--help', 'Display this command-specific help output') do |help|
+              parsed['help'] = true
+            end
+            opts.on('--config CONF', 'Path to the puppet.conf file on disk') do |conf|
+              parsed['config'] = conf
+            end
+          end
+        end
+
+        def run(inputs)
+          config_path = inputs['config']
+
+          # Validate the config path.
+          if config_path
+            errors = FileSystem.validate_file_paths(config_path)
+            return 1 if Errors.handle_with_usage(@logger, errors)
+          end
+
+          # Validate puppet config setting.
+          settings_overrides = {}
+          puppet = Config::Puppet.new(config_path)
+          puppet.load(settings_overrides)
+          return 1 if Errors.handle_with_usage(@logger, puppet.errors)
+
+          # Validate that we are offline
+          return 1 if HttpClient.check_server_online(puppet.settings, @logger)
+
+          # Getting the CRL(s)
+          loader = X509Loader.new(puppet.settings[:cacert], puppet.settings[:cakey], puppet.settings[:cacrl])
+
+          # Take 'em
+          puppet_crl = select_puppet_CRL(loader.crls, loader.key)
+          # Prune 'em
+          prune_CRLs(puppet_crl)
+          # Update 'em
+          update_pruned_CRL(puppet_crl, loader.key)
+          # Write 'em out
+          write_crl_to_file(loader.crls, puppet.settings[:cacrl])
+
+          @logger.inform("Finished pruning Puppet's CRL")
+          return 0
+        end
+
+        def select_puppet_CRL(crl_array,key)
+          puppet_crl = Array.new
+          crl_array.each do |crl|
+            if crl.verify(key)
+              puppet_crl << crl
+            end
+          end
+          return puppet_crl
```

Comment:
You should be able to use the `select` method to do this: https://ruby-doc.org/core-2.7.0/Array.html#method-i-select

We _might_ want to confirm that only one CRL was found, but not totally sure.

Reply:
That's one of the thing that I was a bit unsure about.  When I prune with the `verbose` flag, the Puppet Root CA also shown up so there were 2 CRL that was found.  However, Puppet Root CA revoked list is empty.  I guess I could implement another check for empty revoked list.

Reply:
Nvm, I'm tripping.  That was when I didn't verify the key...

Reply:
heh yeah, at first I thought it would include the root CA, but then I remembered that that uses a different key.